[Zxun2] iP

Zxun2 opened the pull request:
DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/admin/ip-w4.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:


- [X]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)


If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Zxun2 committed:
Add level 1: Greet, Echo, Exit

Zxun2 committed:
Add level 2: Add, List

Zxun2 committed:
Add level 3: Mark, Unmark

Zxun2 committed:
Add level 4: ToDos, Events, Deadlines

Zxun2 committed:
Add A-TextUiTesting: Semi-automated Testing

Zxun2 committed:
Add level 5: Error handling

Zxun2 committed:
Add level 6: Delete

Zxun2 committed:
Add testcases + minor clean up

Zxun2 committed:
Update testcases

Zxun2 committed:
Refactor folder structure + add new functional interface for commands

Zxun2 committed:
Minor clean up

Zxun2 committed:
Add level 7: Save

Zxun2 committed:
Add level 8: Dates

Zxun2 committed:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/entities/Deadline.java
#	src/main/java/entities/Event.java
#	src/main/java/entities/Task.java

Zxun2 committed:
Fix merge conflicts

Zxun2 committed:
Add date command

Zxun2 committed:
Add A-MoreOOP

Zxun2 committed:
Add A-Packages

Zxun2 committed:
Add A-Enums

Zxun2 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Zxun2 committed:
Add A-Gradle and Checkstyle

Zxun2 committed:
Add A-JUnit

Zxun2 committed:
Fix checkstyle violations + refactor executables to receive taskList

Zxun2 committed:
Add A-Jar

Zxun2 committed:
Add JavaDocs

Zxun2 committed:
Merge remote-tracking branch 'origin/branch-A-JavaDoc'

Zxun2 committed:
Add new line between content and meta for JDocs

Zxun2 committed:
Add A-CodingStandard

Zxun2 committed:
Merge branch 'branch-A-CodingStandard'

Zxun2 committed:
Add Level-9

Zxun2 committed:
Format error message for Unmark Command

Zxun2 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/controllers/Command.java
#	src/main/java/controllers/ListCommand.java
#	src/main/java/entities/TaskList.java
#	src/main/java/parser/Parser.java

Zxun2 committed:
Fix merge conflicts

Zxun2 committed:
Update welcome message to include more commands

Zxun2 committed:
Add abstractions to write data after each command

Zxun2 committed:
Reformat DateCommand

Zxun2 committed:
Fix checkstyle violations

Zxun2 committed:
Merge branch 'branch-A-CheckStyle'

Zxun2 committed:
Add Level-10

Zxun2 committed:
Merge branch 'branch-Level-10'

Zxun2 committed:
Update gui

Zxun2 committed:
Add new landing page and router manager

Zxun2 committed:
Add ui image

inas0ng commented on src/main/java/duke/controllers/Command.java:
```diff
@@ -0,0 +1,51 @@
+    protected static final String INVALID_FORMAT_ERROR = "Invalid format.";
+    protected static final Pattern VALID_NUMBER = Pattern.compile("[-+]?\\d+");
+
+    /** The associated type of command **/
```
a small comment, should javadoc of class members be commented in this format instead (single asterisk at the end)?
/** comment here */

inas0ng commented on src/main/java/duke/entities/Deadline.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public boolean activeOn(LocalDate date) {
```
can consider using a more boolean-sounding name here, for e.g., isActiveOn

inas0ng commented on src/main/java/duke/entities/SerializableTask.java:
```diff
@@ -0,0 +1,71 @@
+        case EVENT:
+            task = new Event(description, flags.split(" to "));
+            break;
+        default: task = null;
```
adding a line break behind "default:" may help to improve readability?

inas0ng commented on src/main/java/duke/entities/Todo.java:
```diff
@@ -0,0 +1,41 @@
+public class Todo extends Task {
+
+    /**
+     * Instantiate a todo object.
```
small comment, can consider using "Instantiate**s**" instead of "Instantiate" (applies to the serialize() method javadoc as well)

inas0ng reviewed (commented):
your code is very organised and looks good! it was also nice to learn more about the Matcher class via your code. i've pointed out some minor comments for improvement based on the Java coding standard :smile:

Zxun2 committed:
Update JavaDocs

Zxun2 committed:
Refactor marshal and unmarshal interfaces into generic serializable interface

Zxun2 committed:
Add test stubs and unit tests for task lists

Zxun2 commented:
@inas0ng thanks for reviewing my code!

Zxun2 committed:
Fix coding standard violations

gwynethguo commented on src/main/java/duke/entities/Task.java:
```diff
@@ -0,0 +1,145 @@
+    public static final Pattern DEADLINE =
+            Pattern.compile("^(deadline) (?<description>.+) /by (?<by>.+)$",
+                    Pattern.CASE_INSENSITIVE);
+    public static final Pattern TODO = Pattern.compile("^(todo) (?<description>.+)$", Pattern.CASE_INSENSITIVE);
```
Perhaps can put a line break to follow the format of the code in the lines before.

gwynethguo replied:
Perhaps can put a line break to follow the format of the code in the lines before.

gwynethguo commented on src/main/java/duke/entities/Event.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public boolean isActiveOn(LocalDate date) {
+        return date.isAfter(from) && date.isBefore(to) || date.isEqual(from) || date.equals(to);
```
Perhaps can avoid complicated expressions?

gwynethguo reviewed (commented):
Very good! Your code follows OOP Principles and follows the Java coding standard. Keep up the good work! :D

Zxun2 commented:
@gwynethguo thanks for reviewing my code!😮

Zxun2 committed:
Remove unneeded main function

Zxun2 committed:
Add assertions

There are several methods throughout the codebase with
internal invariants that were assumed but not checked.

Adding assertions verifies the assumptions made about the
program state so that the runtime can verify them.

As a step towards ensuring code correctness in production,
let's add assertions into the codebase where necessary.

Zxun2 committed:
Add A-CodeQuality

There are several code quality violations throughout
the codebase. In some cases, there are code
duplication.

Improving the code quality allows for greater
understandability and readability.

As a step towards increased readability, let's add
fix the obvious violations and extract duplicate
code blocks into separate methods.

Zxun2 committed:
Add A-Streams

Zxun2 committed:
Merge pull request #1 from Zxun2/branch-A-Assertions

Add A-Assertions

Zxun2 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/storage/Storage.java

Zxun2 committed:
Merge pull request #2 from Zxun2/branch-A-CodeQuality

Add A-CodeQuality

Zxun2 committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/duke/entities/TaskList.java
#	src/main/java/duke/storage/Storage.java

Zxun2 committed:
Fix merge conflicts

Zxun2 committed:
Merge pull request #3 from Zxun2/branch-A-Streams

Add A-Streams

Zxun2 committed:
Add Github Workflow

Zxun2 committed:
Remove IO redirection test

Zxun2 committed:
Fix CheckStyle Violations

Zxun2 committed:
Merge pull request #4 from Zxun2/branch-A-CI

Add Github Workflow

Zxun2 committed:
Add C-Undo and Version Control.

A manual process is required to undo a command that
was accidentally entered in. For example, if the
user adds a task, and decides he wants to undo it,
he/she would need to manually enter a delete
command and specifies the task to be deleted. The
same can be said for mark and unmark commands.

With undo, this entire process can be reduced to a
single keyword, "undo". The undo operation is enabled
by keeping track of the last 10 changes. When the
undo command is entered, the most recent changed
state is popped and replaces the current state.

With these changes, it allows for version control to
take place as well. User can now checkout to a
specified version number and save accordingly.

Let's implement the undo function to improve the
user experience and enhance the quality of the application.

Zxun2 committed:
Merge pull request #5 from Zxun2/branch-C-Undo

Add C-Undo and Version Control

Zxun2 committed:
Update main class file in build.gradle

Zxun2 committed:
Update duke messages to better reflect it's personality

Zxun2 committed:
Add more testing

Zxun2 committed:
Add preliminary user guide

Zxun2 committed:
Update release page link

Zxun2 committed:
Update user guide with more details

Zxun2 committed:
Add details for `undo` and `checkout` command in UG

Zxun2 committed:
Fix minor typo in UG

Zxun2 committed:
Update Ui pic

Zxun2 committed:
Fix checkstyle violations

Zxun2 committed:
Fix minor bug

Zxun2 committed:
Update Ui pic (again)

Zxun2 committed:
Update error handling

- Update error message for invalid format
- Improve error handling for list command
- Update header comments

Zxun2 committed:
Update README.md
